### PR TITLE
Better solution for issue #304

### PR DESCRIPTION
**Pull request by erwindevreugd**
This is a better solution for issue #304 
This still shows the parameter dialog for the common parameters when executing a script with parameters
My previous solution/commit prevented that from happening, sorry about that.

### CODE REVIEW

**Comment by HuanhuanSunMSFT** [comment]
@erwindevreugd  Thanks for the PR. It looks good.
Actually, it was intended not to show the parameter dialog if there are no parameters originally. You can find this by right click and you will see no option for Execute with Parameters in the context menu. But we have logged an [issue](https://github.com/Microsoft/poshtools/issues/568) to track this. We should show common parameters even there are no parameters at all. So you actually help fix the part of the issue. 
To fully fix this issue, you may also need to enable showing the execute with Parameters in the context menu, i.e., some changes need to made to code below. Go ahead to make this change if you like. However, this fix itself is good enough to be merged. I can go and finish the aforementioned part.
```
protected override bool ShouldShowCommand(DTE2 dte2)
{
       return base.ShouldShowCommand(dte2) && HasParameters();
}

private bool HasParameters()
{
        _parameterResult = null;
        return ParameterEditorHelper.HasParameters(_adaptersFactory, _textManager, out _paramBlock);
}
```

**Comment by erwindevreugd** [comment]
@HuanhuanSunMSFT Thanks for the tip! I've already made some changes in the direction you've suggested and doing some testing now. I'll send an other merge request later today.

**Comment by erwindevreugd** [comment]
@HuanhuanSunMSFT Hi, I've made some additional changes along the lines that you've suggested. 
You can merge if you want.

File: PowerShellTools/Commands/ExecuteAsScriptCommand.cs [hunk 1]
```diff
@@ -53,12 +61,8 @@ public void Execute(object sender, EventArgs args)
                 return;
 
             Utilities.SaveDirtyFiles();
-            if (ScriptArgs != null && !ScriptArgs.ShouldExecute)
-            {
-                return;
-            }
 
-            launcher.LaunchFile(file, true, ScriptArgs == null ? null : ScriptArgs.ScriptArgs);
+            launcher.LaunchFile(file, true, ScriptArgs != null && ScriptArgs.ShouldExecute ? ScriptArgs.ScriptArgs : null);
```
Comment (HuanhuanSunMSFT) [inline PowerShellTools/Commands/ExecuteAsScriptCommand.cs]: What if user clicks cancel on Parameter editor dialog, which indicates he/she doesn't want to execute the script?

**Comment by erwindevreugd** [comment]
You're right, I noticed that, but hitting cancel on the parameter editor dialog could be interpreted as either cancel script execution or cancel setting parameters but run the script anyway. I didn't want to make any changes without feedback\discussion.

File: PowerShellTools/Commands/ExecuteAsScriptCommand.cs (hunk 1 above)
Comment (erwindevreugd) [inline PowerShellTools/Commands/ExecuteAsScriptCommand.cs]: You're right, I noticed that, but hitting cancel on the parameter editor dialog could be interpreted as either cancel script execution or cancel setting parameters but run the script anyway. I didn't want to make any changes without feedback\discussion.What do you think the cancel button should do?

Comment (HuanhuanSunMSFT) [inline PowerShellTools/Commands/ExecuteAsScriptCommand.cs]: I lean towards the former, cancel script execution. But cancel setting parameters also makes sense. Loop in more people for discussion. @adamdriscoll @jinglou- @AndreSayreMSFT @MrayerMSFT

Comment (erwindevreugd) [inline PowerShellTools/Commands/ExecuteAsScriptCommand.cs]: The biggest issue with the cancel but execute anyway option is likely to be the mandatory parameters. 
And dialogs should always have a true cancel option.

The more I think about it the more I think you're right! :)
But for now I'll wait to hear what the others think.

Comment (MrayerMSFT) [inline PowerShellTools/Commands/ExecuteAsScriptCommand.cs]: I also agree with cancel meaning a cancelling of the script execution. Cancel should (usually) be the user’s way of saying “wait, no, I want to stop what I am doing and go back”.

**Comment by AndreSayreMSFT** [comment]
I'd interpret the user hitting the cancel button as wanting to cancel/back out of executing completely.  Someone can still use the other script execution route to run it without specifying anything.

File: PowerShellTools/Commands/ExecuteAsScriptCommand.cs (hunk 1 above)
Comment (erwindevreugd) [inline PowerShellTools/Commands/ExecuteAsScriptCommand.cs]: Changed clicking cancel on parameter dialog to cancel script execution.

I've also added a suggestion on the issue below which is related to this:
https://github.com/Microsoft/poshtools/issues/568

Also related to this issue:
At the moment the OK button is enabled even when mandatory parameters are not entered allowing the script to be executed anyway. It feels like it should be disabled.

If i could get some feedback that would be great! @adamdriscoll @jinglou- @AndreSayreMSFT  @MrayerMSFT

**Comment by HuanhuanSunMSFT** [comment]
Looks good @erwindevreugd

**Comment by adamdriscoll** [comment]
@HuanhuanSunMSFT Feel free to merge in if you feel this is complete.

**Comment by erwindevreugd** [comment]
It is functional as is but there is still some room for future improvements on these issues:
https://github.com/Microsoft/poshtools/issues/568 (Functional change, still up for discussion)
https://github.com/Microsoft/poshtools/issues/639 (Difficult as current implementation relies on IVsTextBuffer, quick fix would be to disable script execution from Solution Explorer)

If there are any issues with the fix so far please let me know, I'd be happy to follow up!

**Comment by HuanhuanSunMSFT** [comment]
@erwindevreugd  Agree. Also I'm thinking adding parameters saving functionality https://github.com/Microsoft/poshtools/issues/701 to be more efficient.

**Pull request merged**